docs/tutorials/open-source/apache-flink/1_single_feature.qmd

```diff
@@ -0,0 +1,266 @@
+
+
+### Prerequisites
+
+* Docker Compose: This tutorial uses Docker Compose to manage an Apache Kafka environment (including sample data generation) and a Flink cluster (for remote execution). You can [download and install Docker Compose from the official website](https://docs.docker.com/compose/install/).
+* JDK 11 release: Flink requires Java 11.
+* Python 3.9 or 3.10.
+* Follow [the instructions on the Ibis project page](https://ibis-project.org/install) to install the Flink backend for Ibis. For the tutorial below, we assume that you already have the Ibis package correctly installed in your environment.
+* Clone the [example repository](https://github.com/claypotai/ibis-flink-example).
+
+
+### Spinning up the services using Docker Compose
+
+From your project directory, run `docker compose up -d` to create Kafka topics, generate sample data, and launch a Flink cluster in the background.
+
+```{python}
+# | code-fold: true
+!git clone https://github.com/claypotai/ibis-flink-example
+!cd ibis-flink-example && docker compose up -d && cd ..
+```
+
+This should set up a `transaction` topic in the Kafka cluster that contains messages that look like the following:
+
+```
+{
+    "trans_date_trans_time": "2012-02-23 00:10:01", 
+    "cc_num": 4428780000000000000,
+    "merchant": "fraud_Olson, Becker and Koch", 
+    "category": "gas_transport",
+    "amt": 82.55, 
+    "first": "Richard", 
+    "last": "Waters", 
+    "zipcode": "53186",
+    "dob": "1/2/46",
+    "trans_num": "dbf31d83eebdfe96d2fa213df2043586",
+    "is_fraud": 0, 
+    "user_id": 7109464218691269943
+}
+```
+
+
+### Connect to a Flink environment session
+
+We can connect to a Flink environment session by creating a `pyflink.table.TableEnvironment` and passing this to Flink backend’s `connect` method. For this tutorial, we are going to use Flink in streaming mode. 
+
+```{python}
+from pyflink.table import EnvironmentSettings, TableEnvironment
+
+import ibis
+
+env_settings = EnvironmentSettings.in_streaming_mode()
+table_env = TableEnvironment.create(env_settings)
+table_env.get_config().set("parallelism.default", "1")
+connection = ibis.flink.connect(table_env)
+```
+
+The Kafka connector isn’t part of the binary distribution. In order to connect to a Kafka source/sink, we need to [download the JAR file](https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka/3.0.2-1.18/flink-sql-connector-kafka-3.0.2-1.18.jar) and manually add it into the classpath:
+
+```{python}
+# | code-fold: true
+!wget -N https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka/3.0.2-1.18/flink-sql-connector-kafka-3.0.2-1.18.jar
+```
+```{python}
+connection._exec_sql("ADD JAR './flink-sql-connector-kafka-3.0.2-1.18.jar'")
```

Comment:
[Z] Use of a private method here, intended? The blog post will often be referred to as public documentation, i think we need more rigor here.